packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/RangeBondContractForDifferenceFinancialProductLibrary.sol

```diff
@@ -0,0 +1,99 @@
+}
+
+contract RangeBondContractForDifferenceFinancialProductLibrary is
+    ContractForDifferenceFinancialProductLibrary,
+    Lockable
+{
+    using FixedPoint for FixedPoint.Unsigned;
+    using SignedSafeMath for int256;
+
+    struct RangeBondContractForDifferenceParameters {
+        uint256 bondNotional;
+        uint256 highPriceRange;
+        uint256 lowPriceRange;
+    }
+
+    mapping(address => RangeBondContractForDifferenceParameters) public contractForDifferenceParameters;
+
+    function setContractForDifferenceParameters(
+        address contractForDifferenceAddress,
+        uint256 bondNotional,
+        uint256 highPriceRange,
+        uint256 lowPriceRange
+    ) public nonReentrant() {
+        require(
+            ExpiringContractInterface(contractForDifferenceAddress).expirationTimestamp() != 0,
+            "Invalid CFD address"
+        );
+        require(highPriceRange > lowPriceRange, "Invalid bounds");
+
+        RangeBondContractForDifferenceParameters memory params =
+            contractForDifferenceParameters[contractForDifferenceAddress];
+        require(
+            params.bondNotional == 0 && params.lowPriceRange == 0 && params.highPriceRange == 0,
+            "Parameters already set"
+        );
+
+        contractForDifferenceParameters[contractForDifferenceAddress] = RangeBondContractForDifferenceParameters({
+            bondNotional: bondNotional,
+            lowPriceRange: lowPriceRange,
+            highPriceRange: highPriceRange
+        });
+    }
+
+    function computeExpiraryTokensForCollateral(int256 expiryPrice) public view override returns (uint256) {
+        RangeBondContractForDifferenceParameters memory params = contractForDifferenceParameters[msg.sender];
+
+        // A range bond is defined as = Yield Dollar - Put Option + Call option. Numerically this is found using:
+        // N = Notional of bond (100)
+        // P = price of token
+        // T = number of tokens
+        // R1 = Low Price Range
+        // R2 = High Price Range
+        // T = min(N/P,N/R1) + max((N/R2*(P-R2))/P,0)
+        // This represents the value of the long token(range bond holder). This function's method must return a value
+        //  between 0 and 1 to be used as a collateralPerPair that allocates collateral between the short and long tokens.
+        // We can use the value of the long token to compute the relative distribution between long and short CFD tokens
+        // by simply computing longTokenRedeemed using equation above devided by the collateralPerPair of the CFD.
+        // NOTE: the equation's second term could be simplified slightly alebraically to have fewer devided and
+        // multiplications. However this introduces rounding issues. By keeping it in this form this is avoided.
+        // NOTE: the ternary is used over Math.max for the second term to avoid negative numbers.
+
+        uint256 positiveExpiraryPrice = expiryPrice > 0 ? uint256(expiryPrice) : 0;
+
```

Comment:
In light of our offline conversation, here's an alternative implementation of the math (but please let me know if you disagree!).

This really depends on whether our prices are in the form USDCOLLATERAL or COLLATERALUSD. If we assume the more natural COLLATERALUSD, we have to invert everything (in this proposed implementation).

```
FixedPoint.Unsigned memory maxInvertedPrice = FixedPoint.fromUnscaledUint(1).div(FixedPoint.Unsigned(params.lowPriceRange));

FixedPoint.Unsigned memory minInvertedPrice = FixedPoint.fromUnscaledUint(1).div(FixedPoint.Unsigned(params.highPriceRange));

FixedPoint.Unsigned memory invertedPrice = FixedPoint.min(FixedPoint.max(FixedPoint.fromUnscaledUint(1).div(FixedPoint.Unsigned(expiryPrice)), minInvertedPrice), maxInvertedPrice);

return invertedPrice.div(maxInvertedPrice).rawValue;
```

Not sure if this is simpler, but I _think_ this would also work, and a bonus is that it shouldn't require the contract to know anything about the notional.

Thoughts?

Reply:
How did you strip out `N` from `min(N/P,N/R1) + max((N/R2*(P-R2))/P,0)`? Or did you convert `min(N/P,N/R1) + max((N/R2*(P-R2))/P,0)` to another formula?

Reply:
so this is a great meta question and comment. In general, I'm of the opinion that we should try to mimic the "financial" formula as much as possible without introducing abstraction.

Regarding your particular equation, I'm pretty sure this does _not_ work. you _do_ need the notional. From the starting base equation from the Range bond doc of `T = min(N/P,N/R1) + max((N/R2*(P-R2))/P,0)` we can simplify to the equation you posted on slack as `T=min(maxTKN, max(minTKN, N * USDTKN))` which workers fine with an output `T` looking like this:
![image](https://user-images.githubusercontent.com/12886084/120756012-fffe4500-c50e-11eb-8611-7004bfc2aa9d.png)


The formula you posed in your code block does not work correctly (assuming I correctly converted it to normal mathematical notation). it plots like this:

![image](https://user-images.githubusercontent.com/12886084/120756070-17d5c900-c50f-11eb-84b8-44a36765fa55.png)

Reply:
please see the comment below on the update to this.